daemon/emer-aggregate-tally.c

```diff
@@ -0,0 +1,391 @@
+                        payload);
+}
+
+static gchar *
+get_tally_path_from_aggregate_timer (EmerAggregateTally *self,
+                                     guint32             unix_user_id,
+                                     GVariant           *event_id,
+                                     GVariant           *aggregate_key,
+                                     const char         *date)
+{
+  g_autofree gchar *aggregate_key_str = NULL;
+  g_autofree gchar *event_id_str = NULL;
+  g_autofree gchar *filename = NULL;
+
+  aggregate_key_str = g_variant_print (aggregate_key, FALSE);
+  event_id_str = g_variant_print (event_id, FALSE);
+  filename = g_strdup_printf ("%u-%u-%u",
+                              g_str_hash (aggregate_key_str),
+                              unix_user_id,
+                              g_str_hash (event_id_str));
+
+  return g_build_filename (self->persistent_cache_directory,
+                           "aggregate-timers",
+                           date,
+                           filename,
+                           NULL);
+}
+
+static GVariant*
+load_aggregate_timer_from_path (const char *path)
+{
+  g_autoptr(GVariant) variant = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *contents = NULL;
+  gsize size;
+
+  if (!g_file_test (path, G_FILE_TEST_EXISTS))
+    return NULL;
+
+  g_file_get_contents (path, &contents, &size, &error);
+  if (error)
+    return NULL;
+
+  variant = g_variant_new_from_data (G_VARIANT_TYPE (AGGREGATE_TIMER_VARIANT_TYPE),
+                                     contents, size, FALSE, NULL, NULL);
+
+  if (!g_variant_is_normal_form (variant))
+    {
+      g_autoptr(GError) deletion_error = NULL;
+      g_autoptr(GFile) file = NULL;
+
+      g_warning ("Corrupted aggregate timer at %s, deleting...", path);
+
+      file = g_file_new_for_path (path);
+      g_file_delete (file, NULL, &deletion_error);
+      if (deletion_error)
+        g_warning ("Error deleting corrupted tally file %s: %s",
+                   path, deletion_error->message);
+      return NULL;
+    }
+
+  return g_steal_pointer (&variant);
+}
+
+static guint32
+find_previous_aggregate_timer_counter (EmerAggregateTally *self,
+                                       guint32             unix_user_id,
+                                       GVariant           *event_id,
+                                       GVariant           *aggregate_key,
+                                       const char         *date)
+{
+  g_autofree gchar *tally_path = NULL;
+  g_autoptr(GVariant) variant = NULL;
+  guint32 previous_counter = 0;
+
+  tally_path = get_tally_path_from_aggregate_timer (self,
+                                                    unix_user_id,
+                                                    event_id,
+                                                    aggregate_key,
+                                                    date);
+
+  variant = load_aggregate_timer_from_path (tally_path);
```

Comment:
This works, but it loads the entire file from disk just to retrieve a counter. I wonder if it's possible to not load it all?

Reply:
If we are worried about this… can we reconsider using sqlite? :)

Reply:
But what's the alternative (regardless of storage technology)? write a new file with each integer and sum them only when we come to submit the event? Append to the file (which on flash media rewrites the whole file)?

Reply:
Is the file expected to be big (like, tens of kilobytes or larger)? From the looks of the code in this file, it isn’t expected to be big. I wouldn’t worry about it unless profiling shows that this is a performance problem in practice.

I guess any alternative, to be more performant, would have to load only the counter from disk. This code currently loads the entire file into a blob of heap-allocated memory, traverses over all of it checking it’s in normal form, and then indexes the counter from it. A more performant alternative would mmap the file and directly index the counter, or just store the counter in a file by itself.

Reply:
The file is expected to be pretty small, yeah.